// eccrypto-js/dist/cjs/helpers/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ENCRYPT_OP = 'encrypt';
exports.DECRYPT_OP = 'decrypt';
exports.SIGN_OP = 'sign';
exports.VERIFY_OP = 'verify';
exports.AES_LENGTH = 256;
exports.HMAC_LENGTH = 256;
exports.AES_BROWSER_ALGO = 'AES-CBC';
exports.HMAC_BROWSER_ALGO = `SHA-${exports.AES_LENGTH}`;
exports.HMAC_BROWSER = 'HMAC';
exports.SHA256_BROWSER_ALGO = 'SHA-256';
exports.SHA512_BROWSER_ALGO = 'SHA-512';
exports.AES_NODE_ALGO = `aes-${exports.AES_LENGTH}-cbc`;
exports.HMAC_NODE_ALGO = `sha${exports.HMAC_LENGTH}`;
exports.SHA256_NODE_ALGO = 'sha256';
exports.SHA512_NODE_ALGO = 'sha512';
exports.HEX_ENC = 'hex';
exports.UTF8_ENC = 'utf8';
exports.ZERO_LENGTH = 0;
exports.PREFIX_LENGTH = 1;
exports.KEY_LENGTH = 32;
exports.IV_LENGTH = 16;
exports.MAC_LENGTH = 32;
exports.DECOMPRESSED_LENGTH = 64;
exports.PREFIXED_KEY_LENGTH = exports.KEY_LENGTH + exports.PREFIX_LENGTH;
exports.PREFIXED_DECOMPRESSED_LENGTH = exports.DECOMPRESSED_LENGTH + exports.PREFIX_LENGTH;
exports.MAX_KEY_LENGTH = 1024;
exports.MAX_MSG_LENGTH = 32;
exports.EMPTY_BUFFER = Buffer.from(new Uint8Array(0));
exports.EC_GROUP_ORDER = Buffer.from('fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141', exports.HEX_ENC);
exports.ZERO32 = Buffer.alloc(32, 0);
