// @connext/types/dist/src/contracts/misc.js
import { enumify, tidy } from "../utils";
export const EXPECTED_CONTRACT_NAMES_IN_NETWORK_CONTEXT = [
    "ChallengeRegistry",
    "ConditionalTransactionDelegateTarget",
    "DepositApp",
    "IdentityApp",
    "MinimumViableMultisig",
    "MultiAssetMultiPartyCoinTransferInterpreter",
    "ProxyFactory",
    "SingleAssetTwoPartyCoinTransferInterpreter",
    "TimeLockedPassThrough",
    "TwoPartyFixedOutcomeInterpreter",
];
export const singleAssetTwoPartyCoinTransferEncoding = tidy(`tuple(
  address to,
  uint256 amount
)[2]`);
export const multiAssetMultiPartyCoinTransferEncoding = tidy(`tuple(
  address to,
  uint256 amount
)[][]`);
export const OutcomeType = enumify({
    TWO_PARTY_FIXED_OUTCOME: "TWO_PARTY_FIXED_OUTCOME",
    MULTI_ASSET_MULTI_PARTY_COIN_TRANSFER: "MULTI_ASSET_MULTI_PARTY_COIN_TRANSFER",
    SINGLE_ASSET_TWO_PARTY_COIN_TRANSFER: "SINGLE_ASSET_TWO_PARTY_COIN_TRANSFER",
});
export const CommitmentTarget = enumify({
    MULTISIG: "0",
    SET_STATE: "1",
    CANCEL_DISPUTE: "2",
});
