// @connext/types/dist/src/methods.js
import { enumify } from "./utils";
export const MethodNames = enumify({
    chan_create: "chan_create",
    chan_deployStateDepositHolder: "chan_deployStateDepositHolder",
    chan_getAppInstance: "chan_getAppInstance",
    chan_getAppInstances: "chan_getAppInstances",
    chan_getChannelAddresses: "chan_getChannelAddresses",
    chan_getFreeBalanceState: "chan_getFreeBalanceState",
    chan_getProposedAppInstance: "chan_getProposedAppInstance",
    chan_getProposedAppInstances: "chan_getProposedAppInstances",
    chan_getStateChannel: "chan_getStateChannel",
    chan_getStateDepositHolderAddress: "chan_getStateDepositHolderAddress",
    chan_getTokenIndexedFreeBalanceStates: "chan_getTokenIndexedFreeBalanceStates",
    chan_install: "chan_install",
    chan_proposeInstall: "chan_proposeInstall",
    chan_rejectInstall: "chan_rejectInstall",
    chan_sync: "chan_sync",
    chan_takeAction: "chan_takeAction",
    chan_uninstall: "chan_uninstall",
    chan_withdraw: "chan_withdraw",
    chan_withdrawCommitment: "chan_withdrawCommitment",
});
