// @connext/types/dist/src/transfers.js
import { enumify } from "./utils";
import { HashLockTransferAppName, SimpleLinkedTransferAppName, SimpleSignedTransferAppName, SupportedApplicationNames, GenericConditionalTransferAppName, } from "./contracts";
const RequireOnlineAppNames = [
    SupportedApplicationNames.HashLockTransferApp,
];
const AllowOfflineAppNames = [
    SupportedApplicationNames.SimpleSignedTransferApp,
    SupportedApplicationNames.SimpleLinkedTransferApp,
];
export const getTransferTypeFromAppName = (name) => {
    if (RequireOnlineAppNames.includes(name)) {
        return "RequireOnline";
    }
    if (AllowOfflineAppNames.includes(name)) {
        return "AllowOffline";
    }
    return undefined;
};
export const ConditionalTransferTypes = enumify({
    HashLockTransfer: HashLockTransferAppName,
    LinkedTransfer: SimpleLinkedTransferAppName,
    SignedTransfer: SimpleSignedTransferAppName,
});
export const ConditionalTransferAppNames = enumify({
    [HashLockTransferAppName]: HashLockTransferAppName,
    [SimpleLinkedTransferAppName]: SimpleLinkedTransferAppName,
    [SimpleSignedTransferAppName]: SimpleSignedTransferAppName,
    [GenericConditionalTransferAppName]: GenericConditionalTransferAppName,
});
export const LinkedTransferStatus = enumify({
    PENDING: "PENDING",
    COMPLETED: "COMPLETED",
    FAILED: "FAILED",
});
export const HashLockTransferStatus = enumify({
    PENDING: "PENDING",
    EXPIRED: "EXPIRED",
    COMPLETED: "COMPLETED",
    FAILED: "FAILED",
});
export const SignedTransferStatus = enumify({
    PENDING: "PENDING",
    COMPLETED: "COMPLETED",
    FAILED: "FAILED",
});
