// @connext/types/dist/src/watcher.js
export const ChallengeProgressedEvent = "ChallengeProgressedEvent";
export const ChallengeProgressionFailedEvent = "ChallengeProgressionFailedEvent";
export const ChallengeCompletedEvent = "ChallengeCompletedEvent";
export const ChallengeCompletionFailedEvent = "ChallengeCompletionFailedEvent";
export const ChallengeOutcomeSetEvent = "ChallengeOutcomeSetEvent";
export const ChallengeOutcomeFailedEvent = "ChallengeOutcomeFailedEvent";
export const ChallengeCancelledEvent = "ChallengeCancelledEvent";
export const ChallengeCancellationFailedEvent = "ChallengeCancellationFailedEvent";
export const ChallengeUpdatedEvent = "ChallengeUpdatedEvent";
export const StateProgressedEvent = "StateProgressedEvent";
export const WatcherEvents = {
    [ChallengeUpdatedEvent]: ChallengeUpdatedEvent,
    [StateProgressedEvent]: StateProgressedEvent,
    [ChallengeProgressedEvent]: ChallengeProgressedEvent,
    [ChallengeProgressionFailedEvent]: ChallengeProgressionFailedEvent,
    [ChallengeOutcomeSetEvent]: ChallengeOutcomeSetEvent,
    [ChallengeOutcomeFailedEvent]: ChallengeOutcomeFailedEvent,
    [ChallengeCompletedEvent]: ChallengeCompletedEvent,
    [ChallengeCompletionFailedEvent]: ChallengeCompletionFailedEvent,
    [ChallengeCancelledEvent]: ChallengeCancelledEvent,
    [ChallengeCancellationFailedEvent]: ChallengeCancellationFailedEvent,
};
export var StoredAppChallengeStatus;
(function (StoredAppChallengeStatus) {
    StoredAppChallengeStatus[StoredAppChallengeStatus["NO_CHALLENGE"] = 0] = "NO_CHALLENGE";
    StoredAppChallengeStatus[StoredAppChallengeStatus["IN_DISPUTE"] = 1] = "IN_DISPUTE";
    StoredAppChallengeStatus[StoredAppChallengeStatus["IN_ONCHAIN_PROGRESSION"] = 2] = "IN_ONCHAIN_PROGRESSION";
    StoredAppChallengeStatus[StoredAppChallengeStatus["EXPLICITLY_FINALIZED"] = 3] = "EXPLICITLY_FINALIZED";
    StoredAppChallengeStatus[StoredAppChallengeStatus["OUTCOME_SET"] = 4] = "OUTCOME_SET";
    StoredAppChallengeStatus[StoredAppChallengeStatus["CONDITIONAL_SENT"] = 5] = "CONDITIONAL_SENT";
    StoredAppChallengeStatus[StoredAppChallengeStatus["PENDING_TRANSITION"] = 6] = "PENDING_TRANSITION";
})(StoredAppChallengeStatus || (StoredAppChallengeStatus = {}));
