// @connext/utils/dist/channelSigner.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const ethers_1 = require("ethers");
const providers_1 = require("ethers/providers");
const crypto_1 = require("./crypto");
const identifiers_1 = require("./identifiers");
exports.getRandomChannelSigner = (ethProviderUrl) => new ChannelSigner(crypto_1.getRandomPrivateKey(), ethProviderUrl);
class ChannelSigner {
    constructor(privateKey, ethProviderUrl) {
        this.privateKey = privateKey;
        this._ethersType = "Signer";
        this.encrypt = crypto_1.encrypt;
        this.provider = !!ethProviderUrl ? new providers_1.JsonRpcProvider(ethProviderUrl) : undefined;
        this.privateKey = privateKey;
        this.publicKey = crypto_1.getPublicKeyFromPrivateKey(privateKey);
        this.address = crypto_1.getAddressFromPublicKey(this.publicKey);
        this.publicIdentifier = identifiers_1.getPublicIdentifierFromPublicKey(this.publicKey);
    }
    getAddress() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.address;
        });
    }
    decrypt(message) {
        return __awaiter(this, void 0, void 0, function* () {
            return crypto_1.decrypt(message, this.privateKey);
        });
    }
    signMessage(message) {
        return __awaiter(this, void 0, void 0, function* () {
            return crypto_1.signChannelMessage(message, this.privateKey);
        });
    }
    sendTransaction(transaction) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.provider) {
                throw new Error(`ChannelSigner can't send transactions without being connected to a provider`);
            }
            const wallet = new ethers_1.Wallet(this.privateKey, this.provider);
            return wallet.sendTransaction(transaction);
        });
    }
}
exports.ChannelSigner = ChannelSigner;
