// @connext/utils/dist/comparisons.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("ethers/utils");
const bigNumbers_1 = require("./bigNumbers");
exports.notBigNumberish = bigNumbers_1.getBigNumberishError;
exports.notGreaterThan = (value, ceil) => {
    if (exports.notBigNumberish(value)) {
        return exports.notBigNumberish(value);
    }
    return utils_1.bigNumberify(value).gt(utils_1.bigNumberify(ceil))
        ? undefined
        : `Value (${value.toString()}) is not greater than ${ceil.toString()}`;
};
exports.notGreaterThanOrEqualTo = (value, ceil) => {
    if (exports.notBigNumberish(value)) {
        return exports.notBigNumberish(value);
    }
    return utils_1.bigNumberify(value).gte(ceil)
        ? undefined
        : `Value (${value.toString()}) is not greater than or equal to ${ceil.toString()}`;
};
exports.notLessThan = (value, floor) => {
    if (exports.notBigNumberish(value)) {
        return exports.notBigNumberish(value);
    }
    return utils_1.bigNumberify(value).lt(floor)
        ? undefined
        : `Value (${value.toString()}) is not less than ${floor.toString()}`;
};
exports.notLessThanOrEqualTo = (value, floor) => {
    if (exports.notBigNumberish(value)) {
        return exports.notBigNumberish(value);
    }
    return utils_1.bigNumberify(value).lte(floor)
        ? undefined
        : `Value (${value.toString()}) is not less than or equal to ${floor.toString()}`;
};
exports.notPositive = (value) => {
    return exports.notGreaterThanOrEqualTo(value, 0);
};
exports.notNegative = (value) => {
    if (exports.notLessThan(0, value)) {
        return `Value ${value.toString()} is negative.`;
    }
    return undefined;
};
