// @connext/utils/dist/hexStrings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("ethers/utils");
exports.getHexStringError = (value, length) => {
    if (typeof value !== "string") {
        return `Invalid hex string: ${value} is a ${typeof value}, expected a string`;
    }
    if (!value.startsWith("0x")) {
        return `Invalid hex string: ${value} doesn't start with 0x`;
    }
    if (!utils_1.isHexString(value)) {
        return `Invalid hex string: ${value}`;
    }
    if (length && utils_1.hexDataLength(value) !== length) {
        return `Invalid hex string of length ${length}: ${value} is ${utils_1.hexDataLength(value)} bytes long`;
    }
    return undefined;
};
exports.isValidHexString = (value) => !exports.getHexStringError(value);
exports.getAddressError = (value) => {
    try {
        const hexError = exports.getHexStringError(value, 20);
        if (hexError)
            return hexError;
        utils_1.getAddress(value);
        return undefined;
    }
    catch (e) {
        return e.message;
    }
};
exports.isValidAddress = (value) => !exports.getAddressError(value);
exports.getBytes32Error = (value) => {
    const hexStringError = exports.getHexStringError(value, 32);
    if (hexStringError)
        return hexStringError;
    return undefined;
};
exports.isValidBytes32 = (value) => !exports.getBytes32Error(value);
exports.getRandomAddress = () => utils_1.hexlify(utils_1.randomBytes(20));
exports.getRandomBytes32 = () => utils_1.hexlify(utils_1.randomBytes(32));
