// @connext/utils/dist/math.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("ethers/utils");
const constants_1 = require("ethers/constants");
const bigNumbers_1 = require("./bigNumbers");
exports.toWad = (n) => utils_1.parseEther(n.toString());
exports.fromWad = utils_1.formatEther;
exports.weiToToken = (wei, tokenPerEth) => bigNumbers_1.toBN(utils_1.formatEther(exports.toWad(tokenPerEth).mul(wei)).replace(/\.[0-9]*$/, ``));
exports.tokenToWei = (token, tokenPerEth) => exports.toWad(token).div(exports.toWad(tokenPerEth));
exports.maxBN = (lobn) => lobn.reduce((max, current) => (max.gt(current) ? max : current), constants_1.Zero);
exports.minBN = (lobn) => lobn.reduce((min, current) => (min.lt(current) ? min : current), constants_1.MaxUint256);
exports.inverse = (bn) => utils_1.formatEther(exports.toWad(exports.toWad(`1`)).div(exports.toWad(bn)));
exports.calculateExchange = (amount, swapRate) => {
    const [integer, fractional] = swapRate.split(".");
    const safeSwapRate = [integer, (fractional || "0").substring(0, 18)].join(".");
    return utils_1.bigNumberify(utils_1.formatEther(amount.mul(utils_1.parseEther(safeSwapRate))).replace(/\.[0-9]*$/, ""));
};
